File: artisync/Backend/src/main/java/uteq/edu/ec/artisync/service/RolePermissionService.java
package uteq.edu.ec.artisync.service;

import uteq.edu.ec.artisync.dto.request.CreateRoleRequest;
import uteq.edu.ec.artisync.dto.request.UpdateRoleRequest;
import uteq.edu.ec.artisync.dto.response.PermisoResponse;
import uteq.edu.ec.artisync.dto.response.RolResponse;

import java.util.List;

public interface RolePermissionService {

    List<RolResponse> getAllRoles();

    List<PermisoResponse> getAllPermisos();

    List<String> getPermissionsByRole(String roleName);

    void syncPermissions(String roleName, List<String> permissionCodes);

    RolResponse createRole(CreateRoleRequest request);

    RolResponse updateRole(Long idRol, UpdateRoleRequest request);

    void deleteRole(Long idRol);
}

File: artisync/Frontend/src/app/features/administracion/services/role-permission.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { MessageResponse } from '../../../shared/models/common.model';
import {
  RolMatrixResponse, PermisoCatalogResponse, SyncPermissionsRequest,
  CreateRoleRequest, UpdateRoleRequest
} from '../models/admin.model';

@Injectable({
  providedIn: 'root'
})
export class RolePermissionService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/v1/admin/role-permissions`;

  getAllRoles(): Observable<RolMatrixResponse[]> {
    return this.http.get<RolMatrixResponse[]>(`${this.apiUrl}/roles`);
  }

  getAllPermisos(): Observable<PermisoCatalogResponse[]> {
    return this.http.get<PermisoCatalogResponse[]>(`${this.apiUrl}/permisos`);
  }

  getPermissionsByRole(roleName: string): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/${roleName}`);
  }

  syncPermissions(request: SyncPermissionsRequest): Observable<MessageResponse> {
    return this.http.put<MessageResponse>(`${this.apiUrl}/sync`, request);
  }

  createRole(request: CreateRoleRequest): Observable<RolMatrixResponse> {
    return this.http.post<RolMatrixResponse>(`${this.apiUrl}/roles`, request);
  }

  updateRole(idRol: number, request: UpdateRoleRequest): Observable<RolMatrixResponse> {
    return this.http.put<RolMatrixResponse>(`${this.apiUrl}/roles/${idRol}`, request);
  }

  deleteRole(idRol: number): Observable<MessageResponse> {
    return this.http.delete<MessageResponse>(`${this.apiUrl}/roles/${idRol}`);
  }
}
